### PR TITLE
[gce] Fix `NullPointerException`

### LINKED ISSUE #17417: [gce] NPE raised
Found in log, left to investigate the origin : 

```
2024-09-15 09:04:31.718 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.handleCommand()' on 'org.openhab.binding.gce.internal.handler.Ipx800v3Handler@67202fc3': Cannot invoke "java.io.PrintWriter.write(String)" because "this.out" is null
java.lang.NullPointerException: Cannot invoke "java.io.PrintWriter.write(String)" because "this.out" is null
	at org.openhab.binding.gce.internal.handler.Ipx800DeviceConnector.send(Ipx800DeviceConnector.java:64) ~[?:?]
	at org.openhab.binding.gce.internal.model.M2MMessageParser.setOutput(M2MMessageParser.java:110) ~[?:?]
	at org.openhab.binding.gce.internal.handler.Ipx800v3Handler.lambda$8(Ipx800v3Handler.java:333) ~[?:?]
	at java.util.Optional.ifPresent(Optional.java:178) ~[?:?]
	at org.openhab.binding.gce.internal.handler.Ipx800v3Handler.handleCommand(Ipx800v3Handler.java:333) ~[?:?]
	at jdk.internal.reflect.GeneratedMethodAccessor262.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:569) ~[?:?]
	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:147) [bundleFile:?]
	at org.openhab.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
	at jdk.proxy3913.$Proxy4071.handleCommand(Unknown Source) [?:?]
	at org.openhab.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:95) [bundleFile:?]
	at org.openhab.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:49) [bundleFile:?]
	at jdk.internal.reflect.GeneratedMethodAccessor288.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:569) ~[?:?]
	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:147) [bundleFile:?]
	at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:840) [?:?]
```

### PR DESCRIPTION
**Pull request by clinique**
Resolves #17417

### CODE REVIEW

**Comment by lsiepel** [comment]
Should this be backported to 4.2 ?

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 1]
```diff
@@ -60,9 +61,11 @@ public Ipx800DeviceConnector(String hostname, int portNumber, ThingUID uid) {
...
 
     public synchronized void send(String message) {
-        logger.debug("Sending '{}' to Ipx800", message);
-        out.write(message + ENDL);
-        out.flush();
+        output.ifPresentOrElse(out -> {
+            logger.debug("Sending '{}' to Ipx800", message);
+            out.write(message + ENDL);
+            out.flush();
+        }, () -> logger.warn("Tentative to send '{}' while the output stream is closed.", message));
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:68]:
```suggestion
        }, () -> logger.warn("Trying to send '{}' while the output stream is closed.", message));
```

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 2]
```diff
@@ -72,12 +75,15 @@ public synchronized void send(String message) {
      */
     private void connect() throws IOException {
         disconnect();
+
         logger.debug("Connecting {}:{}...", hostname, portNumber);
-        client = new Socket(hostname, portNumber);
+        Socket client = new Socket(hostname, portNumber);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:80]:
```suggestion
        Socket socket = new Socket(hostname, portNumber);
```

Minor but consistent naming woudl be preferred.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java [hunk 3]
```diff
@@ -72,12 +75,15 @@ public synchronized void send(String message) {
...
         client.setSoTimeout(DEFAULT_SOCKET_TIMEOUT_MS);
         client.getInputStream().skip(client.getInputStream().available());
-        in = new BufferedReader(new InputStreamReader(client.getInputStream()));
-        out = new PrintWriter(client.getOutputStream(), true);
+        socket = Optional.of(client);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:83]:
```suggestion
        this.socket = Optional.of(socket);
```

Comes together with the previous comment

**Review by lsiepel — CHANGES_REQUESTED** [review]
Some comments you might want to look in, they are minor though.

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java (hunk 1 above)
Comment (clinique) [reply bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:68]: Corrected

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java (hunk 2 above)
Comment (clinique) [reply bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:80]: Changed

File: bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java (hunk 3 above)
Comment (clinique) [reply bundles/org.openhab.binding.gce/src/main/java/org/openhab/binding/gce/internal/handler/Ipx800DeviceConnector.java:83]: Corrected accordingly.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Pull request merged**